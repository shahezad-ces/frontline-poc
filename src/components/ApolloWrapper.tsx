"use client";
import { ApolloProvider } from "@apollo/client/react";
import client from "@frontline/libs/apolloClient";

export default function ApolloWrapper({ children }: { children: React.ReactNode }) {
  return <ApolloProvider client={client}>{children}</ApolloProvider>;
}