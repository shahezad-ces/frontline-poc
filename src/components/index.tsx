import Categories from "./Categories/Categories";
import PaginationWrapper from "./PaginationWrapper/PaginationWrapper";
import ProductCard from "./ProductCard/ProductCard";
import ProductsFilters from "./ProductsFilters/ProductsFilters";
import AddToCartButton from "./AddToCartButton/AddToCartButton";
import Header from "./Header/Header";

export {
  Categories,
  PaginationWrapper,
  ProductCard,
  ProductsFilters,
  AddToCartButton,
  Header,
};
