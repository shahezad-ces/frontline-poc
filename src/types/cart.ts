import { Product } from "./product";

export type CartItem = Product & { quantity: number };

export type CartContextType = {
  cart: CartItem[];
  addToCart: (product: Product) => void;
  updateQuantity: (quantity: number) => void;
  removeFromCart: (id: number) => void;
  clearCart: () => void;
  totalItems: number;
  totalPrice: number;
  quantity: number;
};
