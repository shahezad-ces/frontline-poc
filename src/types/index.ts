import { Category } from "./category";
import { Product } from "./product";
import { CartItem, CartContextType } from "./cart";

export type { Category, Product, CartItem, CartContextType };
