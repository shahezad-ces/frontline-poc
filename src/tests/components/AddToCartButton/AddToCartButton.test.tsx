import { render, screen, fireEvent, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Product } from "@frontline/types";
import { AddToCartButton } from "@frontline/components";

jest.mock("@frontline/context/CartContext", () => ({
  useCart: jest.fn(),
}));

const mockUseCart = require("@frontline/context/CartContext").useCart;

const mockProduct: Product = {
  id: 1,
  title: "Classic Red Hoodie",
  price: 49.99,
  images: ["https://placehold.co/600x400"],
  category: { id: 1, name: "Clothing", image: "test" },
  description: "A stylish hoodie for everyday wear.",
};

describe("AddToCartButton Component", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });

  it("renders button with correct text", () => {
    mockUseCart.mockReturnValue({ addToCart: jest.fn() });
    render(<AddToCartButton product={mockProduct} />);
    expect(screen.getByTestId("add-to-cart")).toHaveTextContent("Add to Cart");
  });

  it("calls addToCart when clicked", () => {
    const addToCartMock = jest.fn();
    mockUseCart.mockReturnValue({ addToCart: addToCartMock });

    render(<AddToCartButton product={mockProduct} />);
    const button = screen.getByTestId("add-to-cart");

    fireEvent.click(button);
    expect(addToCartMock).toHaveBeenCalledTimes(1);
    expect(addToCartMock).toHaveBeenCalledWith(mockProduct);
  });

  it("shows toast after adding to cart", () => {
    mockUseCart.mockReturnValue({ addToCart: jest.fn() });

    render(<AddToCartButton product={mockProduct} />);
    fireEvent.click(screen.getByTestId("add-to-cart"));

    expect(screen.getByText(/Item added to the cart/i)).toBeInTheDocument();
  });

  it("hides toast after 3 seconds", () => {
    mockUseCart.mockReturnValue({ addToCart: jest.fn() });

    render(<AddToCartButton product={mockProduct} />);
    fireEvent.click(screen.getByTestId("add-to-cart"));

    expect(screen.getByText(/Item added to the cart/i)).toBeInTheDocument();

    act(() => {
      jest.advanceTimersByTime(3000);
    });

    expect(
      screen.queryByText(/Item added to the cart/i)
    ).not.toBeInTheDocument();
  });

  it("matches snapshot", () => {
    mockUseCart.mockReturnValue({ addToCart: jest.fn() });
    const { asFragment } = render(<AddToCartButton product={mockProduct} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
