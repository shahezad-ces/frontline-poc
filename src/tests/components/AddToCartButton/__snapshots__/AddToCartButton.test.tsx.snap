// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AddToCartButton Component matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="rounded-full font-semibold transition-colors duration-200 focus:outline-none bg-blue-800 text-white hover:bg-blue-700 px-5 py-2 text-sm disabled:bg-blue-300 disabled:text-gray-100 disabled:cursor-not-allowed"
    data-testid="add-to-cart"
  >
    Add to Cart
  </button>
</DocumentFragment>
`;
